src/CSV.jl

```diff
@@ -364,141 +385,228 @@ function file(source,
     rowchunkguess = div(rowsguess, N)
-    tapelen = rowchunkguess * 2
     debug && println("parsing using $N threads: $rowchunkguess rows chunked at positions: $ranges")
-    rowsv = Vector{Int}(undef, N)
-    tapesv = Vector{Vector{Vector{UInt64}}}(undef, N)
-    refsv = Vector{Vector{Dict{String, UInt64}}}(undef, N)
-    lastrefsv = Vector{Vector{UInt64}}(undef, N)
+    perthreadrows = Vector{Int}(undef, N)
+    perthreadtapes = Vector{Vector{Vector{UInt64}}}(undef, N)
+    perthreadposlens = Vector{Vector{Vector{UInt64}}}(undef, N)
+    perthreadrefs = Vector{Vector{Dict{String, UInt64}}}(undef, N)
+    perthreadlastrefs = Vector{Vector{UInt64}}(undef, N)
+    perthreadtypecodes = [copy(typecodes) for i = 1:N]
+    perthreadintsentinels = Vector{Vector{Int64}}(undef, N)
     @sync for i = 1:N
 @static if VERSION >= v"1.3-DEV"
         Threads.@spawn begin
             tt = Base.time()
-            trefs = Vector{Dict{String, UInt64}}(undef, ncols)
-            tlastrefs = zeros(UInt64, ncols)
-            ttapes = Vector{UInt64}[Mmap.mmap(Vector{UInt64}, tapelen) for i = 1:ncols]
-            tdatapos = ranges[i]
-            tlen = ranges[i + 1] - (i != N)
-            ret = parsetape(Val(false), ignoreemptylines, ncols, gettypecodes(typemap), ttapes, tapelen, buf, tdatapos, tlen, limit, cmt, EMPTY_POSITIONS, pool, trefs, tlastrefs, rowchunkguess, typecodes, debug, options, true)
+            tl_refs = Vector{Dict{String, UInt64}}(undef, ncols)
+            tl_lastrefs = zeros(UInt64, ncols)
+            tl_tapes, tl_poslens = allocate(rowchunkguess, ncols, typecodes)
+            tl_intsentinels = fill(INT_SENTINEL, ncols)
+            tl_datapos = ranges[i]
+            tl_len = ranges[i + 1] - (i != N)
+            tl_rows, tl_tapes, tl_poslens = parsetape(Val(false), ignoreemptylines, ncols, gettypecodes(typemap), tl_tapes, tl_poslens, buf, tl_datapos, tl_len, limit, cmt, EMPTY_POSITIONS, pool, tl_refs, tl_lastrefs, rowchunkguess, perthreadtypecodes[i], tl_intsentinels, debug, options, true)
             debug && println("thread = $(Threads.threadid()): time for parsing: $(Base.time() - tt)")
-            rowsv[i] = ret[1]
-            tapesv[i] = ret[2]
-            refsv[i] = trefs
-            lastrefsv[i] = tlastrefs
+            perthreadrows[i] = tl_rows
+            perthreadtapes[i] = tl_tapes
+            perthreadposlens[i] = tl_poslens
+            perthreadrefs[i] = tl_refs
+            perthreadlastrefs[i] = tl_lastrefs
+            perthreadintsentinels[i] = tl_intsentinels
         end
 end # @static if VERSION >= v"1.3-DEV"
     end
-    rows = sum(rowsv)
-    tapes = Vector{UInt64}[Mmap.mmap(Vector{UInt64}, rows * 2) for i = 1:ncols]
+    intsentinels = perthreadintsentinels[1]
+    anyintrecode = false
+    # promote typecodes from each thread
+    for col = 1:ncols
+        for i = 1:N
+            @inbounds typecodes[col] = promote_typecode(typecodes[col], perthreadtypecodes[i][col])
```

Comment:
Why apply `@inbounds` only to this line?

Reply:
Applied `@inbounds` to the other line